camera isometric stuff

juliandailey opened the pull request.

juliandailey committed:
camera isometric stuff

Calvinjk commented:
For a PR, try to make sure you take note of what you changed and what reviewers should be looking at!  It looks like you both updated the camera controls as well as the character controller, so this PR isn't just camera isometric stuff!  This info is good to throw in the description of the PR.
![image](https://user-images.githubusercontent.com/8600069/137602884-1b2db1a2-95f8-468d-b3ed-2eb36227528e.png)

Calvinjk commented:
Camera Feedback:

1. Rotation buttons should be Q + E, not Q + R
2. The current rotation method is fine for now, but for https://uss-nebula.atlassian.net/browse/UN-38 a single press of Q or E should move 90 degrees instead of slowly rotating by an amount.
4. With the current switch-case implementation, you can only rotate the camera if it is in an idle state.  Do we want that?  I think its fine, just wondering. 
5. If you change the names of scripts such as CamControl, put that in the description of the PR.  Its totally fine for now, but in the future that might cause dependency issues.
6. For this section of CamControl.cs, I'd just remove all of the old logic you commented out.  GitHub as version control software will let us retrieve that if we want it, currently it is just distracting and code bloat.  I would also remove the comment `// Free target the camera`.  Lastly, remember to remove left and right arrows as camera control options!  Those will be used for player movement, but that can be taken care of in UN-38 if you'd like. 
![image](https://user-images.githubusercontent.com/8600069/137603015-e2256fdd-cb29-4622-a094-28829ff8ba13.png)

Character controller feedback:
1. While Debug.Log() is fantastic for debugging, it will clutter up the console very fast as we put it in more areas of our code.  Can you put a top level flag `bool debugLogs = false` that controls whether or not that script will output debug logs to the console?  That way we can toggle it on a script-by-script basis.
2. I am nitpicking a bit here, but could you rename `Vector3 upMovement` to `Vector3 forwardMovement`?  Its a bit confusing how you swap between forward/up to describe the same thing.  Pick one and stick to it!

Calvinjk reviewed (changes requested):
Good job on the PR!   I did find some things that need to be addressed, but let me know how you want to deal with them and I'll unblock the PR 😄

juliandailey committed:
edits made

- rotation buttons q/r
- rotation available whenever
- rotation by 90 degrees instead of incrementally
- arrow keys no longer available for rotation

Calvinjk commented:
Nice job on the rotation snapping!   It feels pretty good!  If you are interested later on down the line, we can add some lerping to it to smooth out movement a bit.

Some things we will need to tackle:

1. The camera doesn't follow the player.  Further, when you rotate the camera, its rotation point is the player.   The combination of these two things leads to a pretty jarring experience.  I think a good solution would be to always use the IsometricCameraTarget as the rotation point.   That way, whether or not the camera is following the player, the rotation point if always the same place relative to the camera.  
2. The movement controls need to be relative to the camera, not the absolute world space.  If I press `w` I move `up`.   But if I rotate the camera twice in the same direction and now press `w`, I move `down`.  This point is actually moot if we end up using mouse controls, but if we are sticking with WASD it is necessary.  It seems like you tried to do this but there might be a bug somewhere.  I didn't dive into your code, but I can if you need help.
3.  Remember to turn off your debug logs in your final commit!  I don't actually think I was clear enough in my previous feedback here.   Adding in `bool debugLogs = false;` is not a built-in unity feature.   That is the first step, but anywhere you write debug logs, you will also have to wrap in the flag, such as `if (debugLogs) { Debug.LogError("Invalid CameraState"); }`  Lmk if that is unclear or you have questions!
4. This is a nitpick and not at all necessary, but I'm gonna throw it out there to help your coding style since you are in the job search!   Update() and FixedUpdate() tend to become absolutely enormous functions.   Currently you do the Input.GetKeyDown checks directly in Update().  While this is not bad, if we do everything directly in Update(), it gets bloated fast.  A great way to keep logic separated and clean is to create a new function like CheckCameraInputs() and call that from Update().  If you put your logic there, it helps keep Update() clean as well as sectioning off logic in areas it is easier to find and isolate changes.  You did this exact thing well in CharControl.cs!  Even better is to use an InputManager, but we'll talk about that one later.

juliandailey committed:
UN39

- cameratarget now the rotation point
- movement controls now relative to camera using euler
- debug logs now happen with if statements
- update function cleaned up

Calvinjk reviewed (approved):
Alright so setting the IsometricCameraTarget didn't work.  I can explain why in person, but its basically that the way you set up the camera, both the Camera and the CameraTarget are on top of each other (same location).  I actually think this won't work regardless and we will have to do some raycasting to determine the rotation point.  Regardless, I think we should talk about this next week and discuss options.  In the meantime, I made a [new ticket for this](https://uss-nebula.atlassian.net/browse/UN-72) and we'll ignore it in this PR.

Nice work!
